Python/optimizer_analysis.c

```diff
@@ -1,50 +1,1068 @@
+        return NULL;
+    }
+    return func;
+}
+
+
+// 0 on success, anything else is error.
+static int
+ctx_frame_push(
+    _Py_UOpsAbstractInterpContext *ctx,
+    PyCodeObject *co,
+    _Py_UOpsSymType **localsplus_start
+)
+{
+    _Py_UOpsAbstractFrame *frame = frame_new(ctx,
+        co->co_consts, co->co_stacksize,
+        co->co_nlocalsplus,
+        0);
+    if (frame == NULL) {
+        return -1;
+    }
+    if (frame_push(ctx, frame, localsplus_start, co->co_nlocalsplus, 0,
+               co->co_nlocalsplus + co->co_stacksize) < 0) {
+        return -1;
+    }
+    if (frame_initalize(ctx, frame, co->co_nlocalsplus, 0) < 0) {
+        return -1;
+    }
+
+    frame->prev = ctx->frame;
+    ctx->frame = frame;
+
+
+    return 0;
+}
+
+static int
+ctx_frame_pop(
+    _Py_UOpsAbstractInterpContext *ctx
+)
+{
+    _Py_UOpsAbstractFrame *frame = ctx->frame;
+    ctx->frame = frame->prev;
+    assert(ctx->frame != NULL);
+    frame->prev = NULL;
+
+    ctx->water_level = frame->locals;
+    PyMem_Free(frame->sym_consts);
+    ctx->curr_frame_depth--;
+    return 0;
+}
+
+static void
+sym_set_type_from_const(_Py_UOpsSymType *sym, PyObject *obj);
+
+// Steals a reference to const_val
+static _Py_UOpsSymType*
+_Py_UOpsSymType_New(_Py_UOpsAbstractInterpContext *ctx,
+                               PyObject *const_val)
+{
+    _Py_UOpsSymType *self = &ctx->t_arena.arena[ctx->t_arena.ty_curr_number];
+    if (ctx->t_arena.ty_curr_number >= ctx->t_arena.ty_max_number) {
+        OPT_STAT_INC(optimizer_failure_reason_no_memory);
+        DPRINTF(1, "out of space for symbolic expression type\n");
+        return NULL;
+    }
+    ctx->t_arena.ty_curr_number++;
+    self->const_val = NULL;
+    self->types = 0;
+
+    if (const_val != NULL) {
+        Py_INCREF(const_val);
+        sym_set_type_from_const(self, const_val);
+        self->const_val = const_val;
+    }
+
+    return self;
+}
+
+
+static void
+sym_set_type(_Py_UOpsSymType *sym, _Py_UOpsSymExprTypeEnum typ, uint64_t refinement)
+{
+    sym->types |= 1 << typ;
+    if (typ <= MAX_TYPE_WITH_REFINEMENT) {
+        sym->refinement[typ] = refinement;
+    }
+}
+
+// Note: for this, to_sym MUST point to brand new sym.
+static void
+sym_copy_immutable_type_info(_Py_UOpsSymType *from_sym, _Py_UOpsSymType *to_sym)
+{
+    to_sym->types = (from_sym->types & IMMUTABLES);
+    if (to_sym->types) {
+        Py_XSETREF(to_sym->const_val, Py_XNewRef(from_sym->const_val));
+    }
+}
+
+// Steals a reference to obj
+static void
+sym_set_type_from_const(_Py_UOpsSymType *sym, PyObject *obj)
+{
+    PyTypeObject *tp = Py_TYPE(obj);
+
+    if (tp == &PyLong_Type) {
+        sym_set_type(sym, PYLONG_TYPE, 0);
+    }
+    else if (tp == &PyFloat_Type) {
+        sym_set_type(sym, PYFLOAT_TYPE, 0);
+    }
+    else if (tp == &PyUnicode_Type) {
+        sym_set_type(sym, PYUNICODE_TYPE, 0);
+    }
+
+    if (tp->tp_flags & Py_TPFLAGS_MANAGED_DICT) {
```

Comment:
Is this case even used? I put an `abort()` in here and it never got triggered during `make test` (though there are some tests that aren't run that way).